Add support for structured cloning of ErrorInstance objects

cdumez opened the pull request:
#### 7bd6b875a6f7276468c9d900282028fc92f38d59
<pre>
Add support for structured cloning of ErrorInstance objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=257268">https://bugs.webkit.org/show_bug.cgi?id=257268</a>

Reviewed by Yusuke Suzuki and Ryosuke Niwa.

Add support for structured cloning of ErrorInstance objects, per:
- <a href="https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeinternal">https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeinternal</a>

This aligns our behavior with Blink and Gecko.

* LayoutTests/fast/dom/Window/window-postmessage-clone-expected.txt:
* LayoutTests/fast/dom/Window/window-postmessage-clone.html:
* LayoutTests/fast/events/message-port-multi-expected.txt:
* LayoutTests/fast/events/resources/message-port-multi.js:
* LayoutTests/fast/storage/serialized-script-value.html:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/structured-clone.any.worker_81-100-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/structured-clone.any_81-100-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structured-cloning-error-stack-optional.sub.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structuredclone_0-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/window-postmessage.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/semantics/structured-clone/dedicated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/semantics/structured-clone/shared-expected.txt:
* LayoutTests/storage/indexeddb/resources/structured-clone.js:
* LayoutTests/storage/indexeddb/structured-clone-expected.txt:
* LayoutTests/userscripts/window-onerror-for-isolated-world-1-expected.txt:
* LayoutTests/userscripts/window-onerror-for-isolated-world-2-expected.txt:
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::ErrorInstance::finishCreation):
* Source/JavaScriptCore/runtime/ErrorInstance.h:
(JSC::ErrorInstance::createWithoutStackTrace):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):

Canonical link: <a href="https://commits.webkit.org/264552@main">https://commits.webkit.org/264552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbf266e2b5f92061901d65ced24e607e119ac1cb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9612 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8076 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7966 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10232 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8156 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10941 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9197 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9731 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6499 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7288 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14876 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6788 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7622 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7412 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10770 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7530 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7879 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6409 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8126 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7196 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7198 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1856 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1899 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11405 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/8346 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7617 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2005 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1393,24 @@ class CloneSerializer : CloneBase {
                 write(String::fromLatin1(JSC::Yarr::flagsString(regExp->regExp()->flags()).data()));
                 return true;
             }
+            if (auto* errorInstance = jsDynamicCast<ErrorInstance*>(obj)) {
```
According to the spec text, (https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeinternal), we need to obtain `name` via `get` (this is user observable behavior. User can define `name` getter), and check the name via specified one. And after that, we need to run `[[GetOwnProperty]]` to obtain `"message"` and need to check whether message has data property. And it seems that this code is not aligned to the spec's behavior. Can you double-check the spec behavior since they are user observable?

cdumez replied:
> and need to check whether message has data property
That part works and is covered by WPT tests. I know because I got it wrong in my first iteration.

Anyway, I was not aware of the spec so I am not surprised if the implementation doesn't match it closely. I'll look at the spec and align more.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structuredclone_0-expected.txt:
```diff
@@ -26,18 +26,18 @@ PASS ImageData object can be cloned
+PASS TypeError objects can be cloned
+PASS URIError objects can be cloned
+PASS URIError objects from other realms are treated as URIError
+FAIL Cloning a modified Error assert_equals: Checking prototype expected object "TypeError" but got object "URIError"
```
Fixing the issues that Yusuke mentioned fixes this subtest. Will reupload a newer version shortly.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+                auto& vm = m_lexicalGlobalObject->vm();
+                auto scope = DECLARE_THROW_SCOPE(vm);
+                auto errorTypeString = errorInstance->get(m_lexicalGlobalObject, vm.propertyNames->name).toWTFString(m_lexicalGlobalObject);
+                RETURN_IF_EXCEPTION(scope, false);
```
Let's error check twice. `get` can throw error, and can return JSEmpty. And `toWTFString` onto JSEmpty will crash. So,

```
auto errorTypeValue = errorInstance->get(m_lexicalGlobalObject, vm.propertyNames->name);
RETURN_IF_EXCEPTION(scope, false);
auto errorTypeString = errorTypeValue.toWTFString(m_lexicalGlobalObject);
RETURN_IF_EXCEPTION(scope, false);
```

cdumez replied:
Ok.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+
+                String message;
+                PropertyDescriptor messageDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->message, messageDescriptor) && messageDescriptor.isDataDescriptor())
```
Let's do error check after `getOwnPropertyDescriptor`.

cdumez replied:
This code looks correct to me?

If getOwnPropertyDescriptor() threw an exception, it would return false and I already have an exception check after the `if` case.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+
+                unsigned line = 0;
+                PropertyDescriptor lineDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->line, lineDescriptor) && lineDescriptor.isDataDescriptor())
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+
+                unsigned column = 0;
+                PropertyDescriptor columnDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->column, columnDescriptor) && columnDescriptor.isDataDescriptor())
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+
+                String sourceURL;
+                PropertyDescriptor sourceURLDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->sourceURL, sourceURLDescriptor) && sourceURLDescriptor.isDataDescriptor())
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+
+                String stack;
+                PropertyDescriptor stackDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->stack, stackDescriptor) && stackDescriptor.isDataDescriptor())
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+                String message;
+                PropertyDescriptor messageDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->message, messageDescriptor) && messageDescriptor.isDataDescriptor())
+                    message = messageDescriptor.value().toWTFString(m_lexicalGlobalObject);
```
And we need error check after `toWTFString`.

cdumez replied:
There is already an exception check on the next line.

Constellation replied:
Now we have EXCEPTION_ASSERT inside this if-brace, so it is fine :)

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+                unsigned line = 0;
+                PropertyDescriptor lineDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->line, lineDescriptor) && lineDescriptor.isDataDescriptor())
+                    line = lineDescriptor.value().toNumber(m_lexicalGlobalObject);
```
Ditto. `toNumber` requires error check. (Object can be returned, and it can have `valueOf` method).

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+                unsigned column = 0;
+                PropertyDescriptor columnDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->column, columnDescriptor) && columnDescriptor.isDataDescriptor())
+                    column = columnDescriptor.value().toNumber(m_lexicalGlobalObject);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -1390,6 +1433,51 @@ class CloneSerializer : CloneBase {
+                String stack;
+                PropertyDescriptor stackDescriptor;
+                if (errorInstance->getOwnPropertyDescriptor(m_lexicalGlobalObject, vm.propertyNames->stack, stackDescriptor) && stackDescriptor.isDataDescriptor())
+                    stack = stackDescriptor.value().toWTFString(m_lexicalGlobalObject);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -257,6 +259,46 @@ static unsigned typedArrayElementSize(ArrayBufferViewSubtag tag)
+        return ErrorType::TypeError;
+    if (equalLettersIgnoringASCIICase(name, "urierror"_s))
+        return ErrorType::URIError;
+    return ErrorType::Error;
```
We should define these enums separately in this SerializedScriptValue. `ErrorType` enum is not a defined interface in JSC. So this can be changed freely, and it can break serialization since it is relying on this errorType number serialized in persistent storage.

Constellation commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -257,6 +259,46 @@ static unsigned typedArrayElementSize(ArrayBufferViewSubtag tag)
+    case enumToUnderlyingType(ErrorType::URIError):
+        return ErrorType::URIError;
+    default:
+        break;
```
And we should get JSC's ErrorType from defined enum here.

Constellation reviewed (approved):
r=me

cdumez committed:
Add support for structured cloning of ErrorInstance objects
https://bugs.webkit.org/show_bug.cgi?id=257268

Reviewed by Yusuke Suzuki and Ryosuke Niwa.

Add support for structured cloning of ErrorInstance objects, per:
- https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeinternal

This aligns our behavior with Blink and Gecko.

* LayoutTests/fast/dom/Window/window-postmessage-clone-expected.txt:
* LayoutTests/fast/dom/Window/window-postmessage-clone.html:
* LayoutTests/fast/events/message-port-multi-expected.txt:
* LayoutTests/fast/events/resources/message-port-multi.js:
* LayoutTests/fast/storage/serialized-script-value.html:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/structured-clone.any.worker_81-100-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/IndexedDB/structured-clone.any_81-100-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/messagechannel.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structured-cloning-error-stack-optional.sub.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/structuredclone_0-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/window-postmessage.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/semantics/structured-clone/dedicated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/workers/semantics/structured-clone/shared-expected.txt:
* LayoutTests/storage/indexeddb/resources/structured-clone.js:
* LayoutTests/storage/indexeddb/structured-clone-expected.txt:
* LayoutTests/userscripts/window-onerror-for-isolated-world-1-expected.txt:
* LayoutTests/userscripts/window-onerror-for-isolated-world-2-expected.txt:
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::ErrorInstance::finishCreation):
* Source/JavaScriptCore/runtime/ErrorInstance.h:
(JSC::ErrorInstance::createWithoutStackTrace):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):

Canonical link: https://commits.webkit.org/264552@main
